File: challenge_3/elixir/joegotflow83/majority_element/lib/majority_element.ex
defmodule MajorityElement do
  def find_major_ele(input) when is_list(input) do
    input
    |> Enum.reduce(%{}, &count/2)
    |> Map.to_list
    |> determine_majority(length(input))
  end

  def determine_majority([head | tail], n) do
    {num, value} = head
    cond do
      value > (n / 2) ->
        num
      true ->
        determine_majority(tail, n)
    end
  end

  def determine_majority([], _n) do
    "No element is the majority."
  end

  def count(num, acc) do
    Map.update acc, num, 1, &(&1 + 1)
  end
end

File: challenge_3/python/psonghi/src/majority_element.py
def majority_element(array):
    half_time = len(array)/2
    for element in array:
        if array.count(element) > half_time:
            return element

# testing majority_element
array = [2,2,3,7,5,7,7,7,4,7,2,7,4,5,6,7,7,8,6,7,7,8,10,12,29,30,19,10,7,7,7,7,7,7,7,7,7]
print(majority_element(array))
